OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyMapPreviewLogic.cs

```diff
@@ -22,100 +24,39 @@ public class LobbyMapPreviewLogic : ChromeLogic
 		int blinkTick;
 
 		[ObjectCreator.UseCtor]
-		internal LobbyMapPreviewLogic(Widget widget, ModData modData, OrderManager orderManager, LobbyLogic lobby)
+		internal LobbyMapPreviewLogic(Widget widget, ModData modData, OrderManager orderManager, Func<MapPreview> getMap,
+			Action<MapPreviewWidget, MapPreview, MouseInput> onMouseDown, Func<MapPreview, Dictionary<CPos, SpawnOccupant>> getSpawnOccupants)
 		{
 			var mapRepository = modData.Manifest.Get<WebServices>().MapRepository;
 
 			var available = widget.GetOrNull("MAP_AVAILABLE");
 			if (available != null)
 			{
-				available.IsVisible = () => lobby.Map.Status == MapStatus.Available && (!lobby.Map.RulesLoaded || !lobby.Map.InvalidCustomRules);
-
-				var preview = available.Get<MapPreviewWidget>("MAP_PREVIEW");
-				preview.Preview = () => lobby.Map;
-				preview.OnMouseDown = mi => LobbyUtils.SelectSpawnPoint(orderManager, preview, lobby.Map, mi);
-				preview.SpawnOccupants = () => LobbyUtils.GetSpawnOccupants(orderManager.LobbyInfo, lobby.Map);
-
-				var titleLabel = available.GetOrNull<LabelWithTooltipWidget>("MAP_TITLE");
-				if (titleLabel != null)
-				{
-					var font = Game.Renderer.Fonts[titleLabel.Font];
-					var title = new CachedTransform<MapPreview, string>(m => WidgetUtils.TruncateText(m.Title, titleLabel.Bounds.Width, font));
-					titleLabel.GetText = () => title.Update(lobby.Map);
-					titleLabel.GetTooltipText = () => lobby.Map.Title;
-				}
-
-				var typeLabel = available.GetOrNull<LabelWidget>("MAP_TYPE");
-				if (typeLabel != null)
-				{
-					var type = new CachedTransform<MapPreview, string>(m => lobby.Map.Categories.FirstOrDefault() ?? "");
-					typeLabel.GetText = () => type.Update(lobby.Map);
-				}
-
-				var authorLabel = available.GetOrNull<LabelWidget>("MAP_AUTHOR");
-				if (authorLabel != null)
-				{
-					var font = Game.Renderer.Fonts[authorLabel.Font];
-					var author = new CachedTransform<MapPreview, string>(
-						m => WidgetUtils.TruncateText("Created by {0}".F(lobby.Map.Author), authorLabel.Bounds.Width, font));
-					authorLabel.GetText = () => author.Update(lobby.Map);
-				}
+				available.IsVisible = () => getMap().Status == MapStatus.Available && (!getMap().RulesLoaded || !getMap().InvalidCustomRules);
```

Comment:
Our map handling code and parts of our UI are multithreaded, so seeing lines like this makes me uneasy.

We knew from the design of `LobbyLogic` that `lobby.Map` would not change between evaluating e.g. `lobby.Map.Status` and`lobby.Map.RulesLoaded`, but we don't have that guarantee for arbitrary `getMap`s.

Please change these to query `getMap()` once per evaluation, e.g. here:
```csharp
available.IsVisible = () => 
{
	var map = getMap();
	return map.Status == MapStatus.Available && (!map.RulesLoaded || !map.InvalidCustomRules);
};
```